### PR TITLE
[Breaking Change] Allow styling nested tokens with classes in styles

### PR DESCRIPTION
**Pull request by karlhorky**
First of all, thanks for this project, very cool!

I'm just opening an early version of a potential solution for #204 to allow styling things like `.rule` (to override `.atrule`) and `.interpolation-punctuation` (to override `.punctuation`).

This allows to style the `interpolation-punctuation` differently than the normal `punctuation` like this:

Screen Shot 2020-08-24 at 15 09 51

```js
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';

const style = {
  punctuation: {
    color: '#ff0000',
  },
  'punctuation.interpolation-punctuation': {
    color: '#00ff00',
  },
};

const Component = () => {
  const codeString = 'const a = \`\${b}\`;';
  return (
    <SyntaxHighlighter language="javascript" style={style}>
      {codeString}
    </SyntaxHighlighter>
  );
};
```

### TODO

- [x] Recognize selectors with multiple class names and apply styles
- [x] Change script to build stylesheets (`atrule.url` instead of `atrule .token.url`)
- [x] Address test failures

### Why is this a breaking change?

TLDR; This is a breaking change because it changes the way theme keys are authored (the new style is eg. `atrule.url` instead of `atrule .token.url`).

The old style (`atrule .token.url`) wasn't working before anyway (see #204), so this is technically not breaking the behavior of themes:

- they didn't work before
- without changes, they will also not work with the new version

However, the benefit is, if the themes are changed to the new format, colors that never worked before will start working in the new version. The change required is minimal:

```diff
-"atrule .token.url": {
+"atrule.url": {
    "color": "#9cdcfe"
},
-"atrule .token.url .token.function": {
+"atrule.url.function": {
    "color": "#dcdcaa"
},
-"atrule .token.url .token.punctuation": {
+"atrule.url.punctuation": {
    "color": "#d4d4d4"
},
```

**Why was it like that?** The [`refractor` stylesheets script](https://github.com/react-syntax-highlighter/react-syntax-highlighter/blob/master/scripts/build-stylesheets-refractor.js) did a simple conversion on this CSS without much processing:

```CSS
.token.atrule .token.url {
	color: #9cdcfe;
}

.token.atrule .token.url .token.function {
	color: #dcdcaa;
}

.token.atrule .token.url .token.punctuation {
	color: #d4d4d4;
}
```

It was transformed into this JS (note the extra ` .token` bits):

```js
"atrule .token.url": {
    "color": "#9cdcfe"
},
"atrule .token.url .token.function": {
    "color": "#dcdcaa"
},
"atrule .token.url .token.punctuation": {
    "color": "#d4d4d4"
},
```

The script has now been updated in this PR, so the new correct format will be generated:

```js
"atrule.url": {
    "color": "#9cdcfe"
},
"atrule.url.function": {
    "color": "#dcdcaa"
},
"atrule.url.punctuation": {
    "color": "#d4d4d4"
},
```

**Why the new format?**

The flat structure of the DOM in `react-syntax-highlighter` (see https://github.com/react-syntax-highlighter/react-syntax-highlighter/pull/305#issuecomment-683288877) means it makes more sense to author each key similar to a CSS selector with multiple chained classes (eg. `atrule.url`). This is different from the current format, which was a simple conversion of the Prism theme selectors without processing.

It also makes it more efficient and less ambiguous to match the theme keys to the node `className`: https://github.com/karlhorky/react-syntax-highlighter/blob/5cf65a0bf828c0f02cee033cbb895670ac1a4e88/src/create-element.js#L5-L39

### Original issue

See problem reproduction here: https://github.com/react-syntax-highlighter/react-syntax-highlighter/issues/204#issuecomment-679094884

Ref https://github.com/react-syntax-highlighter/react-syntax-highlighter/issues/204

### CODE REVIEW

**Comment by simmerer** [comment]
Glad to see you've tackled this... I had a similar nested-token-styling issue in a different project that relies on RSH, and hadn't yet attempted a solution. I'll take a look as soon as I can.

**Comment by karlhorky** [comment]
Great, thanks!

Since I've worked a bunch on https://github.com/PrismJS/prism-themes/pull/103 recently, I guess there is a weakness in this approach:

It doesn't cover the case of features requiring descendant combinator CSS (the space), such as `.token.imports .token.maybe-class-name`:

https://github.com/PrismJS/prism/pull/2533

Or... wait, are classes of parents and descendants combined when using `react-syntax-highlighter` / `refractor`?

So that they would actually look like `.imports.maybe-class-name` at the end? That may work...

**Comment by karlhorky** [comment]
Ah I guess there's a DOM element missing on the `react-syntax-highlighter` side for this:

### `prism`

Screen Shot 2020-08-28 at 17 20 48


### `react-syntax-highlighter`

Screen Shot 2020-08-28 at 17 20 26


### `refractor`

`refractor` does already return this structure from `refractor.highlight` (demo: https://runkit.com/embed/mfigg7b3rcb5):

Screen Shot 2020-08-28 at 17 46 12

**Comment by karlhorky** [comment]
@simmerer if I can ask, is there a requirement by `react-syntax-highlighter` to **NOT** include this extra DOM element for whatever reason?

Because otherwise, I could also take a look at adding in these extra containers...

**Comment by simmerer** [comment]
@karlhorky Not that I know of. @conorhastings is the original creator of this package (I just stepped in to assist), so he might have an opinion.

In my understanding, Prism (or Highlight) themes should work the same in `react-syntax-highlighter` as if you were using Prism (or Highlight) directly, which means the DOM should provide all of the expected structure. If `refractor` provides those elements, we probably should as well.

We'd be super grateful for a PR from you!

**Comment by karlhorky** [comment]
Ok sounds good. I'll take a look at this over the next days, see if I can figure out why that DOM structure is not being reproduced, and try my hand at fixing it.

**Comment by karlhorky** [comment]
It looks like the flattening behavior comes from a `flattenCodeTree` function written in 2017 by @conorhastings - https://github.com/react-syntax-highlighter/react-syntax-highlighter/pull/64

It was originally introduced to avoid some undesirable wrapping behavior (see #63)

Maybe we can work around the wrapping problems differently.

**Comment by karlhorky** [comment]
Does the `wrapLines` option actually work? Can't find an example where it actually wraps to the container width properly.

**Comment by karlhorky** [comment]
Oh I guess `wrapLines` does not mean what I (and others) think it does.

Looking at the original issue https://github.com/react-syntax-highlighter/react-syntax-highlighter/issues/44, it's about wrapping each line **IN A CONTAINER ELEMENT**, not wrapping when the code size exceeds the width.

Maybe the option name should be changed to something less ambiguous...

**Comment by karlhorky** [comment]
Ok @simmerer I gave it another shot, this time styling all permutations and array subsets (power sets) of the classnames on a node.

It avoids having to rework the DOM structure (which is that way for a reason), and should be pretty performant (usually adds up to only a few milliseconds to style a whole document).

Let me know what you think, and if it's acceptable, we can try to land this (I suppose we'll need tests? or to adapt existing ones?)

**Comment by karlhorky** [comment]
Oh actually, wait - why am I trying to do this in an algorithmic way when I can just be declarative about it? I don't think the number of class names will be greater than 3. Even with 4, adding all combinations would still be faster.

This is way better and more performant:

```diff
- function recursivePermute(arr, permutation = []) {
-   return arr.length === 0
-     ? [permutation]
-     : arr.flatMap((item, i, arr) =>
-         recursivePermute(
-           arr.filter((_, j) => j !== i),
-           [...permutation, item]
-         )
-       );
- }
- 
- const permutations = {};
- function permute(arr) {
-   const key = arr.join(".");
-   if (!permutations[key]) {
-     permutations[key] = recursivePermute(arr);
-   }
-   return permutations[key];
- }
- 
- function getSubArrays(arr) {
-   if (arr.length === 1) return [arr];
-   else {
-     const subarr = getSubArrays(arr.slice(1));
-     return subarr.concat(
-       subarr.map((el) => el.concat(arr[0])),
-       [[arr[0]]]
-     );
-   }
- }
+ // Get all possible permutations of all power sets
+ //
+ // Super simple, non-algorithmic solution since the
+ // number of class names will not be greater than 3
+ function powerSetPermutations(arr) {
+   if (arr.length === 0) return [];
+   if (arr.length === 1) return arr;
+   if (arr.length === 2) {
+     return [
+       arr[0],
+       arr[1],
+       arr[0, 1].join('.'),
+       arr[1, 0].join('.'),
+     ]
+   }
+   if (arr.length >= 3) {
+     // Currently does not support more than 3 class names
+     return [
+       arr[0],
+       arr[1],
+       arr[2],
+       arr[0, 1].join('.'),
+       arr[0, 2].join('.'),
+       arr[1, 0].join('.'),
+       arr[1, 2].join('.'),
+       arr[2, 0].join('.'),
+       arr[2, 1].join('.'),
+       arr[0, 1, 2].join('.'),
+       arr[0, 2, 1].join('.'),
+       arr[1, 0, 2].join('.'),
+       arr[1, 2, 0].join('.'),
+       arr[2, 0, 1].join('.'),
+       arr[2, 1, 0].join('.'),
+     ]
+   }
+ }

const classNameCombinations = {};
function getClassNameCombinations(classNames) {
  if (classNames.length === 0) return [];
  if (classNames.length === 1) return classNames;
  const key = classNames.join(".");
  if (!classNameCombinations[key]) {
-     classNameCombinations[key] = getSubArrays(classNames)
-       .flatMap((classNames) => classNames.length > 1 ? permute(classNames) : [classNames])
-       // Less specific (fewer class names) to more specific (more class names)
-       .sort((a, b) => (a.length < b.length ? -1 : 1))
-       .map((arr) => arr.join("."));
+     classNameCombinations[key] = powerSetPermutations(classNames);
  }
  return classNameCombinations[key];
}
```

**Comment by karlhorky** [comment]
Ok, this is looking much cleaner now.

One breaking change that this will require is that class names in the styles no longer have the spaces that they currently have. I have noted this above in the description.

**Comment by karlhorky** [comment]
Alright I've also fixed the build script now, seems to be working.

**Comment by karlhorky** [comment]
Ok, I've also ran the tests and found some other incompatibilities that I've now fixed.

The only 2 tests that are failing also fail on the `master` branch:

Screen Shot 2020-08-30 at 16 00 11

So this PR should be ready to go!

**Comment by simmerer** [comment]
Great work here @karlhorky. I should be able to take a look tomorrow... thanks for your patience. 

In regards to `wrapLines`, I've got a PR here which introduces a `wrapLongLines` prop that does what everyone thought `wrapLines` did. Would love your thoughts on that: #309

**Comment by karlhorky** [comment]
Cool! Wrote a few comments. Not sure about the overall approach and if there aren't any edge cases with using `pre-wrap`, but seems like it could be legit!

**Comment by karlhorky** [comment]
@simmerer I guess you haven't had a chance to take a look at this one yet, right? Any estimate on when you'll be able to take a look?

File: src/create-element.js [hunk 1]
```diff
@@ -1,7 +1,59 @@
 import React from 'react';
 
+// Get all possible permutations of all power sets
+//
+// Super simple, non-algorithmic solution since the
+// number of class names will not be greater than 3
```
Comment (simmerer) [inline src/create-element.js:6]: Nice move here.

File: src/create-element.js [hunk 2]
```diff
@@ -37,28 +89,48 @@ export default function createElement({
...
+      props = {
+        ...properties,
+        className: createClassNameString(properties.className)
+      };
+    } else {
+      const allStylesheetSelectors = [
+        ...new Set(
+          Object.keys(stylesheet)
+            .map(selector => selector.split('.'))
+            .flat()
```
Comment (simmerer) [inline src/create-element.js:105]: I've previously run into issues with ES2019 syntax (`.flat()` in particular) in this project, since we're still building on Node v10: https://github.com/react-syntax-highlighter/react-syntax-highlighter/pull/249

No particular reason why we shouldn't update to use new hotness. I just haven't undertaken that yet, and a quick switch to node 12 in my local environment introduced a different test failure for the async tests.

**Review by simmerer — CHANGES_REQUESTED** [review]
Sorry for taking so long to look this over! Generally looks great.

My one minor request is to trade out `.flat()` for a spread/concat operation, since this project still builds against node 10 for now.

File: src/create-element.js (hunk 2 above)
Comment (karlhorky) [reply src/create-element.js]: Oh, ok! I've done it with `.reduce()` instead here:

[`91fc5d8` (#305)](https://github.com/react-syntax-highlighter/react-syntax-highlighter/pull/305/commits/91fc5d875d98c37889f9079dc44637ee7db55047)

**Review by simmerer — APPROVED** [review]

**Comment by karlhorky** [comment]
Thanks for the merge @simmerer!

Super exciting that this is out now in `15.0.1` (sorry for missing the fixes on the demo themes)!

I'll give this a shot soon!

**Comment by simmerer** [comment]
No worries. `npm publish` (which runs `npm run build`) is still frequently the point at which I realize I haven't run `npm run build` to update the demos. :)

**Comment by karlhorky** [comment]
Ah, @simmerer would it be possible to update to the latest `prism-themes@1.4.1` version too? (similar to your commit 7e26d31cf689721d3fe45532397f4854f9ea2c17 ?)

I did a pull request to improve the accuracy of the VS Code Dark theme: https://github.com/PrismJS/prism-themes/pull/103#issuecomment-703639184

**Comment by simmerer** [comment]
I'll have a PR up shortly!

**Comment by simmerer** [comment]
🎉 Released in `15.1.0`: https://github.com/react-syntax-highlighter/react-syntax-highlighter/releases/tag/v15.1.0

**Comment by karlhorky** [comment]
Awesome, thanks for that!
